Add file extension for Starlark

mcuadros opened the pull request:
## Description
This PR's adds `.star` as an extension to the Starlark language. The `.star` extension is the default extension for Starlark scripts not related to Bazel, this is growing every day since many projects are embedding the language in their own projects.

This extension is used in all official interpreters in the examples and as well in the [testdata](https://github.com/bazelbuild/starlark/tree/master/test_suite/testdata).

Also is the official extension recommended by many opensource projects:
- https://docs.drone.io/pipeline/scripting/starlark/#context
- https://github.com/qri-io/qri/tree/master/startf/testdata
- https://github.com/mcuadros/ascode/tree/master/_examples

 This was discarded initially at https://github.com/github/linguist/pull/4759#discussion_r362961419

The amount of `.star` files currently is [30k](https://github.com/search?q=extension%3Astar+NOT+nothack&type=Code), since the PR instrudicing the language the usage has growth a 50%.
## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- Feel free to remove whole sections, not points within the sections, that do not apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] **I am associating a language with a new file extension.**
  - [x] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      - https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Astar+NOT+nothack
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/dockhippie/nodejs/blob/a04a35ef94418d559b434aa4834b9e71960fd42f/.drone.star
    - Sample license(s):
https://github.com/dockhippie/nodejs/blob/master/LICENSE
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

pchaigno commented:
I don't think this pull request addresses the concerns expressed in #4759. The issue was not with the in-the-wild usage, but with the presence of many other `.star` files. All these files will be recognized as Starlark files with the pull request as is.

mcuadros commented:
Sorry, but if I am not mistaken, in linguist are many extension collisions,
that are resolved with the classifier. For example `.md`.

So all this files are not going to be identified as Starlark since a sample
is provided for the classifier. I am wrong?

Usually this can be a problem because this consume more resources than just
check the extension. But given the number of .star is slow, I don't believe
this is going to be a problem.





On Sun, Apr 19, 2020, 9:47 PM Paul Chaignon <notifications@github.com>
wrote:

> I don't think this pull request addresses the concerns expressed in #4759
> <https://github.com/github/linguist/pull/4759>. The issue was not with
> the in-the-wild usage, but with the presence of many other .star files.
> All these files will be recognized as Starlark files with the pull request
> as is.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/github/linguist/pull/4840#issuecomment-616214309>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMAB6XTV43Z3ETYLJCPWRDRNNIOZANCNFSM4MF2JEBA>
> .
>

pchaigno commented:
> Sorry, but if I am not mistaken, in linguist are many extension collisions,
that are resolved with the classifier. For example `.md`.
>
> So all this files are not going to be identified as Starlark since a sample
is provided for the classifier. I am wrong?

That is correct, but requires us to also have support in Linguist for the conflicting languages. In the case of `.md`, Linguist knows both GCC Machine Description and Markdown and can use the classifier (and other, more precise strategies) to distinguish between the two. So for `.star`, we'll need to identity the conflicting languages and add them to Linguist if they meet usage requirements.

mcuadros commented:
Well but if the other languages are not identified will remain the same,
but the Starlark files will be identified.

Even if doesn't work like this, the amount files that aren't Starlark it's
minimal. And Starlark is growing everyday.

Please reconsider it.


On Sun, Apr 19, 2020, 10:52 PM Paul Chaignon <notifications@github.com>
wrote:

> Sorry, but if I am not mistaken, in linguist are many extension collisions,
> that are resolved with the classifier. For example .md.
>
> So all this files are not going to be identified as Starlark since a sample
> is provided for the classifier. I am wrong?
>
> That is correct, but requires us to also have support in Linguist for the
> conflicting languages. In the case of .md, Linguist knows both GCC
> Machine Description and Markdown and can use the classifier (and other,
> more precise strategies) to distinguish between the two. So for .star,
> we'll need to identity the conflicting languages and add them to Linguist
> if they meet usage requirements.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/github/linguist/pull/4840#issuecomment-616223392>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMAB6SJLQKC67N5CDGSAQ3RNNQCRANCNFSM4MF2JEBA>
> .
>

fkorotkov commented:
@pchaigno I want to bump this issue. What will require for us to "identity the conflicting languages"?

I made a [search query to look for recently indexed *.star files](https://github.com/search?o=desc&q=extension%3Astar+filename%3A%2A.star+-path%3Apkg+-path%3Ago+-path%3Ainfra&s=indexed&type=Code), checked a few first pages and it seems majority of cases are Startlark cases. The rest is some data files or files without a particular pattern.

mahmoudimus commented:
bump! seconding @fkorotkov. /cc @lildude  @pchaigno - any way we can get this looked at?

lildude commented:
> bump! seconding @fkorotkov. /cc @lildude @pchaigno - any way we can get this looked at?

Things are still in the same place as it was... you still need to identify the other users of the `.star` extension. From a very [quick look](https://github.com/search?q=extension%3Astar+loop_), there's one particular use of `.star` files easily identifiable that is pretty active with nearly 8000 files and each file appears to occur only once per repo making it a clear example that qualifies for inclusion else it would be incorrectly identified. I have no idea what language it's written in, but it appears to be related to machine learning from the few READMEs I've looked at so it will be noticed very quickly if it's misidentified.

You don't have to identify _every_ user of the `.star` extension, but an attempt needs to be made to at least identify the most popular (probably just the one I've found), and then add them (it) as part of this PR (I've re-opened it).

fkorotkov commented:
Just for a reference we at [Cirrus CI](https://github.com/marketplace/cirrus-ci) started using Starlark along side YAML for allowing to [programmatically configure CI builds](https://cirrus-ci.org/guide/programming-tasks/). So there will be more `.cirrus.star` and `lib.star` files on GitHub in the near future. 😅

rohansingh commented:
@lildude Just so I understand, if we go through and identify all the conflicting languages that meet usage requirements, do they need to be added as new languages in this same PR?

lildude commented:
@rohansingh I think you may have missed the last sentence in my last response:

> You don't have to identify _every_ user of the `.star` extension, but an attempt needs to be made to at least identify the most popular (probably just the one I've found), and then add them (it) as part of this PR (I've re-opened it).

rohansingh commented:
Sorry, reading comprehension failure. Anyway, sounds good, I'll try to get to this today.

lildude reviewed (changes requested):
Marking this as still needing changes

lildude commented:
> You don't have to identify _every_ user of the `.star` extension, but an attempt needs to be made to at least identify the most popular (probably just the one I've found), and then add them (it) as part of this PR (I've re-opened it).

I've done this. I've identified the other most commonly used language is [Self-defining Text Archive and Retrieval](https://en.wikipedia.org/wiki/Self-defining_Text_Archive_and_Retrieval) commonly referred to as STAR.

I've pushed changes to your repo to add support to this PR.

Alhadis reviewed (commented):
That [`pro-dna.nmr.star`](https://github.com/github/linguist/blob/6ee517cf21dec5c00a2d0893a6ff11e9512aeaaf/samples/STAR/pro-dna.nmr.star) sample is _huge_ (8,613 lines / 145 KBs). Sure we can't trim it down a bit? I noticed that a lot of sections repeat the same data.

lildude commented:
> That [`pro-dna.nmr.star`](https://github.com/github/linguist/blob/6ee517cf21dec5c00a2d0893a6ff11e9512aeaaf/samples/STAR/pro-dna.nmr.star) sample is _huge_ (8,613 lines / 145 KBs). Sure we can't trim it down a bit? I noticed that a lot of sections repeat the same data.

🤦 And I point this out myself so often 😆 I'll get the ✂️ out.

lildude commented:
@Alhadis replaced it with a smaller sample.
